L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,463 @@
+// verifier implementations can compose their functionality with existing
+// implementations of this interface, such as HostnameVerifier, by delegating
+// to an instance of that class.
+class CertificateVerifier {
+ public:
+  ~CertificateVerifier();
+
+  // Verifies a connection request. The check on each internal verifier could be
+  // either synchronous or asynchronous, and we will need to use return value to
+  // know.
+  //
+  // request: the verification information associated with this request
+  // callback: This will only take effect if the verifier is asynchronous.
+  //           The function that gRPC will invoke when the verifier has already
+  //           completed its asynchronous check. Callers can use this function
+  //           to perform any additional checks. The input parameter of the
+  //           std::function indicates the status of the verifier check.
+  // sync_status: This will only be useful if the verifier is synchronous.
+  //              The status of the verifier as it has already done it's
+  //              synchronous check.
+  // return: return true if executed synchronously, otherwise return false
+  bool Verify(TlsCustomVerificationCheckRequest* request,
+              std::function<void(Status)> callback,
+              Status* sync_status);
+
+  // Cancels a verification request previously started via Verify().
+  // Used when the connection attempt times out or is cancelled while an async
+  // verification request is pending.
+  //
+  // request: the verification information associated with this request
+  void Cancel(TlsCustomVerificationCheckRequest* request);
+};
+
+// A CertificateVerifier that doesn't perform any additional checks other than
+// certificate verification, if specified.
+// Note: using this solely without any other authentication mechanisms on the
+// peer identity will leave your applications to the MITM(Man-In-The-Middle)
+// attacks. Users should avoid doing so in production environments.
+class NoOpCertificateVerifier : public CertificateVerifier {
+ public:
+  NoOpCertificateVerifier();
+};
+
+// A CertificateVerifier that will perform hostname verification, to see if the
+// target name set from the client side matches the identity information
+// specified on the server's certificate.
+class HostNameCertificateVerifier : public CertificateVerifier {
+ public:
+  HostNameCertificateVerifier();
+};
+```
+
+#### Custom Chain Building
+```c++
+class CustomChainBuilderInterface {
+public:
+ // TODO(gtcooke94) - more detail here on output specifics.
+ // The output chain should consist of DER encoded certs ordered in the same way
+ // that Open/BoringSSL build chains in `X509_verify_cert` including the root of
+ // the trusted chain.
+ virtual absl::StatusOr<std::vector<absl::string_view>>
+ BuildAndVerifyChain(const std::vector<absl::string_view>& peer_cert_chain_der)
+ = 0;
+ virtual ~CustomChainBuilderInterface() = 0;
+}
```

Comment:
Similar to what we do above for the HostnameVerifier, maybe we can add an "example" for the SPIFFE federation case?

Reply:
Added below the paragraph, PTAL

Reply:
Can we have some kind of a lead-in from the previous paragraph? For example,

"""
One use case for the custom chain building feature is to enable SPIFFE federation (TODO: ref the SPIFFE spec), where the set of root certificates to use is determined based on the SPIFFE trust domain in the peer's leaf certificate.
"""

Reply:
PTAL